File: examples/12_redos/safe.ts
import { McpServer } from "@modelcontextprotocol/sdk/server/mcp.js";
import { z } from "zod";

const server = new McpServer({ name: "safe-redos-demo", version: "1.0.0" });

// SAFE: Rewritten without nested quantifiers; linear time complexity
// Also enforce input length limits before running regex
const EMAIL_RE = /^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$/;
const SLUG_RE  = /^[a-z0-9]+(?:-[a-z0-9]+)*$/;

server.tool("validate_email", { email: z.string().max(254) }, async ({ email }) => {
  const valid = EMAIL_RE.test(email);
  return { content: [{ type: "text", text: valid ? "Valid" : "Invalid" }] };
});

server.tool("validate_slug", { slug: z.string().max(128) }, async ({ slug }) => {
  const match = SLUG_RE.test(slug);
  return { content: [{ type: "text", text: match ? "Match" : "No match" }] };
});

File: examples/12_redos/safe.py
"""
Safe: Input length limit + linear-time regex alternatives.
"""
import re
from mcp.server.fastmcp import FastMCP

mcp = FastMCP("Safe Server")

MAX_INPUT_LEN = 1000

@mcp.tool()
def validate_input(user_input: str) -> bool:
    """Validate input — length-limited, no nested quantifiers."""
    if len(user_input) > MAX_INPUT_LEN:
        return False
    # SAFE: anchored, possessive-style rewrite — no nested quantifiers
    pattern = re.compile(r"^[a-z]+$")
    return bool(pattern.match(user_input))

@mcp.tool()
def validate_email(email: str) -> bool:
    """Validate email — linear-time pattern."""
    if len(email) > 254:  # RFC 5321 max
        return False
    # SAFE: no nested quantifiers
    return bool(re.match(r"^[a-zA-Z0-9._%+\-]+@[a-zA-Z0-9.\-]+\.[a-zA-Z]{2,}$", email))
